sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,314 @@
+        }
+
+        mutex_unlock(&_rb_mutex);
+    }
+
+    /* initially trigger chained async calls (chained via "_gap_event_cb") */
+    if (mutex_trylock(&_send_mutex)) {
+        if (!_sending_in_progress) {
+            _send_from_ringbuffer(true);
+        }
+        else {
+            /* there is an active call chain, we don't must not initiate a new one */
+            mutex_unlock(&_send_mutex);
+        }
+    }
+}
+
+void stdio_init(void)
+{
+    #ifndef STDIO_NIMBLE_DISABLE_UART
+        _init_stdio_uart(&_stdin_isrpipe);
+    #endif
+
+    ringbuffer_init(&_rb, _rb_mem, RINGBUFFER_SIZE);
+}
+
+ssize_t stdio_read(void *buffer, size_t count)
+{
+    return (ssize_t)isrpipe_read(&_stdin_isrpipe, buffer, count);
+}
+
+ssize_t stdio_write(const void* buffer, size_t len)
+{
+    #ifndef STDIO_NIMBLE_DISABLE_UART
+        uart_write(STDIO_UART_DEV, (const uint8_t*) buffer, len);
+    #endif
+
+    _stdio_nimble_write(buffer, len);
+
+    return len;
+}
+
+/* is gonna be called by auto_init */
+void stdio_nimble_init(void)
+{
+    int rc = 0;
+
+    /* verify and add our custom services */
+    rc = ble_gatts_count_cfg(_gatt_svr_svcs);
+    assert(rc == 0);
+    rc = ble_gatts_add_svcs(_gatt_svr_svcs);
+    assert(rc == 0);
+
+    /* reload the GATT server to link our added services */
+    ble_gatts_start();
+
+    /* register gap event listener */
+    rc = ble_gap_event_listener_register(&_gap_event_listener, _gap_event_cb, NULL);
+    assert(rc == 0);
+
+    /* fix compilation error when using DEVELHELP=0 */
+    (void) rc;
+}
+
+/* there should be always only one active call chain */
+/* prevent multiple chains by "_send_mutex" and "_sending_in_progress" */
+static void _send_from_ringbuffer(bool mutex_locked)
+{
+    if (!mutex_locked) {
+        mutex_lock(&_send_mutex);
+    }
+
+    mutex_lock(&_rb_mutex);
+
+    if (ringbuffer_empty(&_rb)) {
+        _sending_in_progress = false;
+
+        mutex_unlock(&_rb_mutex);
+        mutex_unlock(&_send_mutex);
+
+        return;
+    }
+
+    int len_read = ringbuffer_get(&_rb, _rb_read_buf, _current_max_read);
+
+    mutex_unlock(&_rb_mutex);
+
+    int rc = 1;
+    if (_send_stdio) {
+        if (_conn_handle != 0 && _val_handle_stdout != 0) {
+
+            struct os_mbuf *om = ble_hs_mbuf_from_flat(_rb_read_buf, len_read);
+
+            if (om != NULL) {
+                _sending_in_progress = true;
+                rc = ble_gattc_indicate_custom(_conn_handle, _val_handle_stdout, om);
+            }
+        }
+    }
+
+    if (rc != 0) {
+        _sending_in_progress = false;
+    }
+
+    mutex_unlock(&_send_mutex);
+}
+
+static void _clear_ringbuffer(void)
+{
+    mutex_lock(&_rb_mutex);
+    ringbuffer_init(&_rb, _rb_mem, RINGBUFFER_SIZE);
+    mutex_unlock(&_rb_mutex);
+}
+
+static int _gap_event_cb(struct ble_gap_event *event, void *arg)
+{
+    (void) arg;
+
+    switch (event->type) {
+
+        case BLE_GAP_EVENT_CONNECT:
+            _clear_ringbuffer();
```

Comment:
why clear it? Wouldn't it be nicer to get all the node's output that was produced prio to the connect event? This way one could even see the boot messages (provided the buffer space is configured accordingly).

Reply:
The problem I see here is that this is somewhat a strange behaviour. You also wouldn't expect to get old output from your old session when you start a new ssh session on a remote machine. In my opinion it could limit the use cases of this module, because people don't want to get the old output. Either because it's irrelevant output of a previous session or even worse something that should be kept secret and is now sent to someone else. In addition, even if not clearing the buffer, the boot message would be sent only on the first connection. But instead one could think about sending a welcome message after connecting.